contrib/mesos/pkg/scheduler/podtask/resources.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podtask
+
+import (
+	"math"
+
+	"github.com/gogo/protobuf/proto"
+	mesos "github.com/mesos/mesos-go/mesosproto"
+	"github.com/mesos/mesos-go/mesosutil"
+)
+
+func procureScalarResources(name string, value float64, roles []string, offer *mesos.Offer) []*mesos.Resource {
+	roleCaps := map[string]float64{}
+	res := filterResources(offer.Resources, isScalar, hasName(name))
+
+	for _, r := range res {
+		role := starredRole(r.GetRole())
+		roleCaps[role] += r.GetScalar().GetValue()
+	}
+
+	resources := []*mesos.Resource{}
+	for _, r := range roles {
+		if capacity := roleCaps[r]; capacity > 0 {
+			consumed := math.Min(capacity, value)
+			value -= consumed
+
+			res := mesosutil.NewScalarResource(name, consumed)
+			res.Role = protoNilString(r)
+			resources = append(resources, res)
+
+			if value == 0.0 {
```

Comment:
golang implements IEEE 754 which guarantees that 0.0 is an actual (not rounded) number.

IEEE 754 also guarantees that the calculation of arithmetic is equal to the real operation plus rounding (for each single arithmetic operation). We have to dig into math.Min to understand whether there is a chance to get rounding errors which harm us.

In any case, using `value <= 0.0` might be a good idea. Moreover, using e.g. `value < 0.000001` might sense. That's less than a byte and irrelevant for cpu shares.